### DISCUSSION TITLE
Downgrading from Flux v2.0.0-rc1

### DISCUSSION THREAD

**Discussion by andreipantelimon** [fluxcd/flux2#3828] [Q&A] [answered]
Hey,
Is there any procedure from downgrading from Flux v2.0.0-rc1 to the previous version v0.41.2 without replacing resources?
After bootstrapping with the previous version, we are running into issues with CRD's.
`CustomResourceDefinition/gitrepositories.source.toolkit.fluxcd.io dry-run failed, reason: Invalid, error: CustomResourceDefinition.apiextensions.k8s.io "gitrepositories.source.toolkit.fluxcd.io" is invalid: status.storedVersions[1]: Invalid value: "v1": must appear in spec.versions`
Thanks

**Comment by kingdonb** [comment]
I've heard it said that there is no ready process for downgrading from the v1 CRD. This is unfortunate but it is the state of things.

If you need to downgrade from Flux with v1 CRDs, you may need to uninstall and reinstall Flux. I know that we should have all our Flux resources in Git, since this is GitOps, but I know many people run `flux install` and `kubectl apply` their resources to the cluster instead. If you do this, you might need to backup your Flux resources so you can restore them after. In general, if people are following GitOps and using Git as their single source of truth, uninstall/reinstall should be very easy, since you'll have a commit you can revert where the upgrade was done.

If there is an error with your upgrade, can we ask what cluster provider you're using? I'm more interested in how you arrived at this error and how to help you upgrade productively, rather than trying to downgrade which is difficult and basically unsupported. We can only effectively support the latest version.

What version of Kubernetes and what provider are you on? (Is it Kubernetes 1.27.0?) We did become aware of an issue with that version that manifested as an occasional `dry-run failed` error, this could be what you're experiencing. If so, please try upgrading to the next K8s patch version. The K8s maintainers took our report right away and released a new version quickly.

* https://github.com/kubernetes/kubernetes/issues/117260

This sounds like it might be the same issue you're experiencing, but it's hard to know for sure without more details about your setup.

**Reply by kingdonb** [reply, marked as answer]
If you need to downgrade Flux, we don't have an official instructions for this, but here is the process assuming you have bootstrapped and are following the directions normally:

* Uninstall Flux with `flux uninstall` - make sure you're using the same CLI version that you have installed on the cluster
  * This will first shut down all Flux controllers
  * It will remove all Flux Custom Resources and delete their finalizers so they are garbage collected by Kubernetes
  * It will leave any Flux workloads intact, (since Flux controllers are down there's nothing to garbage collect/prune workloads)
  * It will then remove the Flux CRDs, the last trace of Flux on the cluster
  * (If the uninstall hangs, at this point you may have mismatched the CLI version with the version of Flux installed)
* Revert the upgrade commit
* Revert any upgrades to the API version in your Flux manifests that you might have made according to the upgrade instructions
* Install the old Flux CLI version that you wish to revert to
* Bootstrap Flux again with the desired CLI version

Hope this helps!
